### PR TITLE
Laws 2

### PR DESCRIPTION
**Pull request by edmundnoble**
This depends on #1934. 
It adds laws for comonads, cobind, and traverse.

It also adds all laws for homomorphisms.

### CODE REVIEW

**Comment by edmundnoble** [comment]
@NeQuissimus Just so you know in advance, I'll add a microsite page to explain the laws as soon as I've integrated them with testz, which is the next step I want to take.

File: laws/shared/src/main/scala/scalaz/laws/HomomorphismLaws.scala [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+package laws
+
+import scala.{ Nothing, Unit }
+
+import data._, tc._, Scalaz._
+
+object HomomorphismLaws {
+  // To be a monoid homomorphism, a function must satisfy both `monoidIdentity`
+  // and `semigroupAppend`.
+  def monoidIdentity[A, B, T](f: A => B)(assert: (B, B) => T)(implicit A: Monoid[A], B: Monoid[B]): T =
```
Comment (Jacoby6000) [inline laws/shared/src/main/scala/scalaz/laws/HomomorphismLaws.scala]: If this is testing homomorphisms, then shouldn't it be testing in both directions?  This only verifies half of the homomorphism, unless I'm misunderstanding something.

(Same question applies to other assertions in this file)

Comment (edmundnoble) [reply laws/shared/src/main/scala/scalaz/laws/HomomorphismLaws.scala]: I don't understand; homomorphisms go in one direction.

Comment (edmundnoble) [reply laws/shared/src/main/scala/scalaz/laws/HomomorphismLaws.scala]: Though that reminds me, it would be nice to have isomorphism laws too.

Comment (Jacoby6000) [reply laws/shared/src/main/scala/scalaz/laws/HomomorphismLaws.scala]: Oh yeah my bad. I got mixed up. Pretend that I was telling you to write `Iso` laws.

File: project/ScalazBuild.scala [hunk 2]
```diff
@@ -22,7 +22,6 @@ object Scalaz {
     "-opt-warnings",
     "-Xlint:_,-type-parameter-shadow",
     "-Xsource:2.13",
-    "-Ywarn-dead-code",
```
Comment (alexknvl) [inline project/ScalazBuild.scala]: :heart:

**Pull request merged**